File: content/tracks/dsa-python/concepts/02-sliding-window/sliding-window_test.py
import sys
from stub import longest_unique_substring, max_sum_subarray, min_window_substring

passed = 0
failed = 0


def test(name: str, condition: bool, detail: str = "") -> None:
    global passed, failed
    if condition:
        print(f"  PASS: {name}")
        passed += 1
    else:
        msg = f" — {detail}" if detail else ""
        print(f"  FAIL: {name}{msg}")
        failed += 1


# longest_unique_substring
r = longest_unique_substring("abcabcbb")
test("unique: abcabcbb", r == 3, f"got {r}")
r = longest_unique_substring("bbbbb")
test("unique: all same", r == 1, f"got {r}")
r = longest_unique_substring("pwwkew")
test("unique: pwwkew", r == 3, f"got {r}")
r = longest_unique_substring("")
test("unique: empty", r == 0, f"got {r}")
r = longest_unique_substring("a")
test("unique: single char", r == 1, f"got {r}")
r = longest_unique_substring("abcdefg")
test("unique: all distinct", r == 7, f"got {r}")
r = longest_unique_substring("dvdf")
test("unique: dvdf", r == 3, f"got {r}")

# max_sum_subarray
r = max_sum_subarray([2, 1, 5, 1, 3, 2], 3)
test("max_sum: k=3", r == 9, f"got {r}")
r = max_sum_subarray([2, 3, 4, 1, 5], 2)
test("max_sum: k=2", r == 7, f"got {r}")
r = max_sum_subarray([1, 2, 3, 4, 5], 5)
test("max_sum: k=length", r == 15, f"got {r}")
r = max_sum_subarray([5, 1, 1, 1, 1], 1)
test("max_sum: k=1", r == 5, f"got {r}")
r = max_sum_subarray([-1, -2, -3, -4], 2)
test("max_sum: all negative", r == -3, f"got {r}")

# min_window_substring
r = min_window_substring("ADOBECODEBANC", "ABC")
test("min_window: ADOBECODEBANC/ABC", r == "BANC", f"got {r!r}")
r = min_window_substring("a", "a")
test("min_window: exact match", r == "a", f"got {r!r}")
r = min_window_substring("a", "b")
test("min_window: no match", r == "", f"got {r!r}")
r = min_window_substring("", "a")
test("min_window: empty s", r == "", f"got {r!r}")
r = min_window_substring("aa", "aa")
test("min_window: duplicate in t", r == "aa", f"got {r!r}")
r = min_window_substring("cabwefgewcwaefgcf", "cae")
test("min_window: longer string", r == "cwae", f"got {r!r}")

print(f"\n{passed} passed, {failed} failed")
if failed > 0:
    sys.exit(1)

File: content/tracks/dsa-javascript/concepts/02-sliding-window/sliding-window_test.js
const assert = require("assert");
const { longestUniqueSubstring, maxSumSubarray, minWindowSubstring } = require("./stub");

let passed = 0, failed = 0;

function test(name, fn) {
  try {
    fn();
    console.log(`  PASS: ${name}`);
    passed++;
  } catch (e) {
    console.log(`  FAIL: ${name} — ${e.message}`);
    failed++;
  }
}

// longestUniqueSubstring
test("longestUniqueSubstring: classic abcabcbb", () => {
  assert.strictEqual(longestUniqueSubstring("abcabcbb"), 3);
});
test("longestUniqueSubstring: all same character", () => {
  assert.strictEqual(longestUniqueSubstring("bbbbb"), 1);
});
test("longestUniqueSubstring: pwwkew", () => {
  assert.strictEqual(longestUniqueSubstring("pwwkew"), 3);
});
test("longestUniqueSubstring: empty string", () => {
  assert.strictEqual(longestUniqueSubstring(""), 0);
});
test("longestUniqueSubstring: all unique", () => {
  assert.strictEqual(longestUniqueSubstring("abcdef"), 6);
});
test("longestUniqueSubstring: single character", () => {
  assert.strictEqual(longestUniqueSubstring("z"), 1);
});
test("longestUniqueSubstring: duplicate not adjacent (dvdf)", () => {
  assert.strictEqual(longestUniqueSubstring("dvdf"), 3);
});
test("longestUniqueSubstring: two chars alternating", () => {
  assert.strictEqual(longestUniqueSubstring("abababab"), 2);
});

// maxSumSubarray
test("maxSumSubarray: window in middle", () => {
  assert.strictEqual(maxSumSubarray([2, 1, 5, 1, 3, 2], 3), 9);
});
test("maxSumSubarray: window of 2", () => {
  assert.strictEqual(maxSumSubarray([2, 3, 4, 1, 5], 2), 7);
});
test("maxSumSubarray: window equals array length", () => {
  assert.strictEqual(maxSumSubarray([1, 1, 1, 1], 4), 4);
});
test("maxSumSubarray: window of 1", () => {
  assert.strictEqual(maxSumSubarray([3, 1, 4, 1, 5], 1), 5);
});
test("maxSumSubarray: negative numbers", () => {
  assert.strictEqual(maxSumSubarray([-1, -2, -3, -4], 2), -3);
});
test("maxSumSubarray: best at start", () => {
  assert.strictEqual(maxSumSubarray([9, 1, 1, 1], 2), 10);
});
test("maxSumSubarray: best at end", () => {
  assert.strictEqual(maxSumSubarray([1, 1, 1, 9], 2), 10);
});

// minWindowSubstring
test("minWindowSubstring: classic ADOBECODEBANC", () => {
  assert.strictEqual(minWindowSubstring("ADOBECODEBANC", "ABC"), "BANC");
});
test("minWindowSubstring: exact match", () => {
  assert.strictEqual(minWindowSubstring("a", "a"), "a");
});
test("minWindowSubstring: no valid window", () => {
  assert.strictEqual(minWindowSubstring("a", "b"), "");
});
test("minWindowSubstring: t has duplicates", () => {
  assert.strictEqual(minWindowSubstring("aa", "aa"), "aa");
});
test("minWindowSubstring: t longer than s", () => {
  assert.strictEqual(minWindowSubstring("a", "aa"), "");
});
test("minWindowSubstring: window is full string", () => {
  assert.strictEqual(minWindowSubstring("abc", "abc"), "abc");
});
test("minWindowSubstring: multiple valid windows picks shortest", () => {
  // "a...c" covers ABC but "BANC" is shorter
  assert.strictEqual(minWindowSubstring("ADOBECODEBANC", "ABC"), "BANC");
});
test("minWindowSubstring: single char t repeated in s", () => {
  assert.strictEqual(minWindowSubstring("aabc", "b"), "b");
});

console.log(`\n${passed} passed, ${failed} failed`);
if (failed > 0) process.exit(1);
